src/executor/read_only_runtime_host.rs

```diff
@@ -0,0 +1,389 @@
+                } else {
+                    req.key()
+                };
+
+                if let Some(key) = key {
+                    assert!(key > requested_key);
+                }
+
+                // The next key can be either the one passed by the user or one key in the current
+                // pending storage changes that has been inserted during the execution.
+                // As such, find the "next key" in the list of overlay changes.
+                // TODO: not optimized in terms of searching time ; should really be a BTreeMap or something
+                let in_overlay = self
+                    .inner
+                    .top_trie_changes
+                    .iter()
+                    .map(|(k, v)| (k, v.is_some()))
+                    .filter(|(k, _)| &***k > requested_key)
+                    .min_by_key(|(k, _)| *k);
+
+                let outcome = match (key, in_overlay) {
+                    (Some(a), Some((b, true))) if a <= &b[..] => Some(a),
+                    (Some(a), Some((b, false))) if a < &b[..] => Some(a),
+                    (Some(a), Some((b, false))) => {
+                        debug_assert!(a >= &b[..]);
+                        debug_assert_ne!(&b[..], requested_key);
+
+                        // The next key according to the parent storage has been erased earlier in
+                        // the block execution. It is necessary to ask the user again, this time
+                        // for the key after the one that has been erased.
+                        // This `clone()` is necessary, as `b` borrows from
+                        // `self.inner.top_trie_changes`.
+                        let key_overwrite = Some(b.clone());
+                        self.inner.vm = host::HostVm::ExternalStorageNextKey(req);
+                        return RuntimeHostVm::NextKey(NextKey {
+                            inner: self.inner,
+                            key_overwrite,
+                        });
+                    }
+                    (Some(a), Some((b, true))) => {
+                        debug_assert!(a >= &b[..]);
+                        Some(&b[..])
+                    }
+
+                    (Some(a), None) => Some(a),
+                    (None, Some((b, _))) => Some(&b[..]),
+                    (None, None) => None,
+                };
+
+                self.inner.vm = req.resume(outcome.as_ref().map(|v| &v[..]));
+            }
+
+            // We only create a `NextKey` if the state is one of the above.
+            _ => unreachable!(),
+        };
+
+        self.inner.run()
+    }
+}
+
+/// Implementation detail of the execution. Shared by all the variants of [`RuntimeHostVm`]
+/// other than [`RuntimeHostVm::Finished`].
+struct Inner {
+    /// Virtual machine running the call.
+    vm: host::HostVm,
+
+    /// Pending changes to the top storage trie that this execution performs.
+    top_trie_changes: HashMap<Vec<u8>, Option<Vec<u8>>, fnv::FnvBuildHasher>,
```

Comment:
```suggestion
```

This simplifies a lot of the rest of the code. I'm not going to add suggestions everywhere.